### PR TITLE
refactor ActiveRecord's #become by removing not needed code

### PR DESCRIPTION
**Pull request by dcrec1**
This refactor throw some warnings when running the tests:

```
warning: instance variable @changed_attributes not initialized
```

If this is a problem, I can change the line to be like this:

```
  became.instance_variable_set("@changed_attributes", @changed_attributes) if defined?(@changed_attributes)
```

### CODE REVIEW

**Comment by kaspth** [comment]
Sorry, but this is cosmetic change. Thanks :heart:

**Comment by rafaelfranca** [comment]
The code you proposed is exactly what it was before https://github.com/rails/rails/commit/5f6370a81bd013b801f3e4842ffd466d756d127d, so either the test is wrong, or the bug doesn't exist.

**Comment by dcrec1** [comment]
@rafaelfranca maybe the bug existed and the test is ok, but some changes in another parts of the code now prevent the bug from happening.

@kaspth actually there is less code, less behaviour and consequently the code is easier to understand and consequently easier to change. I remember some years ago the core team was interested in keeping the code easier to maintain, sorry if this changed, I didn't know.

**Comment by rafaelfranca** [comment]
Check with the code without warning if the test passes.

**Comment by kaspth** [comment]
@dcrec1 be nice now :smile:

**Comment by rafaelfranca** [comment]
Less and simpler code is fine, just cosmetic change is not.

**Comment by rafaelfranca** [comment]
As I suspected the tests are failing now. So that unnecessary code is just to make sure our code has no warnings and I think it is fine. Do you have other way how to write that code so we don't have the warnings and it is simpler?

**Comment by kaspth** [comment]
This might do:

```ruby
became.instance_variable_set(:@changed_attributes, defined?(@changed_attributes) ? @changed_attributes : {})
```

**Comment by dcrec1** [comment]
I have another way but it actually changes some behaviour. The change would be removing the `freeze` from this [line](https://github.com/rails/rails/blob/70ac072976c8cc6f013f0df3777e54ccae3f4f8c/activerecord/lib/active_record/attribute_methods/dirty.rb#L62) and changing the committed line to:

    became.instance_variable_set("@changed_attributes", changed_attributes)

In case this `freeze` could be removed, I'll create a new pull request with all the changes in one commit.

**Comment by rafaelfranca** [comment]
If removing the freeze it will work means that we are mutating an object that should not be mutated, so I'd say it may be a dangerous change.

**Comment by dcrec1** [comment]
I've tracked this change to this [commit](https://github.com/rails/rails/commit/877ea78). Couldn't really understand why the `freeze` was added, it's clear to you? Seems the author tried to force some method to be called, but didn't understand why.

**Comment by rafaelfranca** [comment]
The reason was exactly to force people to not change that object directly and use the right methods because direct changes of it do not reflect the expected behavior.

**Comment by dcrec1** [comment]
Anyway, after better analyzing and understanding the code, I changed the line to be like:

    became.instance_variable_set("@changed_attributes", attributes_changed_by_setter)

This new way doesn't throw warning neither removes the freeze and I think it's better than `#changed_attributes` because it uses what's already changed instead of re-doing the process.

Is this ok?

**Comment by rafaelfranca** [comment]
But will not this calculate the changed attributes for the case where `@changed_attributes` is not defined (the case that trows the exception)?

**Comment by dcrec1** [comment]
In this case the original [`#changed_attributes`](https://github.com/rails/rails/blob/master/activemodel/lib/active_model/dirty.rb#L169) method will be called, he does basically what the code before the refactor did.

**Comment by rafaelfranca** [comment]
I see. Yeah that would work, but it will be so much more indirection and the code would be hard to follow. Not sure if it is an improvement over the current code.

To avoid one `if` and one `||` we are requiring how want to understand this method to look in another file and understand that `attributes_changed_by_setter` is an alias to `attributes_changed` and it will not call the `ActiveRecord::Dirty` version because we called the alias, not the overridden method.

**Comment by dcrec1** [comment]
But for understanding what this method is doing it shouldn't be necessary to deep into the stack. I think that in this case the variable is well named and explains what it represents and that is very clear that the code is passing the changed attributes to the new class. I agree that the underlying code could be better, maybe `changed_attributes` shouldn't be overridden and maybe that `freeze` could be removed, as the original method doesn't use it and the new method, depending of the condition, doesn't use it neither. I'd happy to tried to improve this code if you agree.

**Comment by rafaelfranca** [comment]
I'll defer this to @sgrif since he wrote the original code.

**Comment by rafaelfranca** [comment]
BTW, your solution looks good to me, so if he don't have anything against I'll merge it using `attributes_changed_by_setter`

**Comment by dcrec1** [comment]
Sorry I sent a wrong commit, will send again.

**Comment by rafaelfranca** [comment]
Thank. Lets wait @sgrif review
